### PR TITLE
UPSTREAM: 14537: Add Persistent Volume Provisioner Controller

### PR DESCRIPTION
**Pull request by markturansky**
Adds the dynamic provisioning framework w/ testable HostPath provisioner.

@jsafrane's 3 provisioners (EBS,Cinder,GCE) are not yet merged upstream.  They will require separate patches into Origin. 

@liggitt @pweil-

### CODE REVIEW

**Comment by jsafrane** [comment]
* There is something wrong in initialization of recyclers/deleters - they are not called on AWS, GCE and OpenStack. When a clam is deleted, appropriate persistent volume stays 'Released' forever.

* Also, when a claim is deleted, it takes quite a long time (several minutes) to move it from Bound to Released.

**Comment by markturansky** [comment]
@jsafrane I added the ebs/gce/cinder plugins to the list of recyclable plugins (w/ comment stating why). Those were the missing bits in Origin.  HostPath was already in the recyclable list, so that implementation worked for me.

**Comment by markturansky** [comment]
The default sync period is 10 minutes, hence the long wait.  This can be changed via CLI flag.  It is tedious to wait during development, but a production system won't care that it waited up to 10m before the volume becomes Released.

**Comment by jsafrane** [comment]
I confirm, with the latest patches my provisioners delete volumes (sometimes after 10 minutes though).

File: pkg/cmd/server/start/start_master.go [hunk 1]
```diff
@@ -558,6 +558,7 @@ func startControllers(oc *origin.MasterConfig, kc *kubernetes.MasterConfig) erro
 		kc.RunEndpointController()
 		kc.RunNamespaceController()
 		kc.RunPersistentVolumeClaimBinder()
+		kc.RunPersistentVolumeProvisioner(pvKClient)
```
Comment (pweil-) [inline pkg/cmd/server/start/start_master.go]: @liggitt @deads2k - are we going to want separate identities for the provisioner and recycler?  

@markturansky - if so this needs the same treatment as the PVController - service account, role, policy

Comment (deads2k) [inline pkg/cmd/server/start/start_master.go]: > @liggitt @deads2k - are we going to want separate identities for the provisioner and recycler?

Yeah, I think they'll end up with different identities.  I would expect more power for the provisioner.

Comment (markturansky) [inline pkg/cmd/server/start/start_master.go]: The provisioners we have today only call a cloud provider's API.

Comment (markturansky) [inline pkg/cmd/server/start/start_master.go]: I'd like to not add an account at this point.

The next iteration of provisioning (which includes recycling operations) contains a "service account name" bit, which will be the account to use for that specific plugin.  Users will be able to mix and match according to the needs of the plugin.

Comment (deads2k) [inline pkg/cmd/server/start/start_master.go]: > The provisioners we have today only call a cloud provider's API.

Then that definitely seems to be a case for a separate SA that has no rights.

Comment (pweil-) [inline pkg/cmd/server/start/start_master.go]: Not sure if you're referring to the pods that get launched or not but the controller itself needs permissions.  From what it looks like this controller needs permissions for CRUD on PVs, PV status, and PVCs which is different from the recycler which also includes permissions for pods and events.  Is that an incorrect summary?

Comment (markturansky) [inline pkg/cmd/server/start/start_master.go]: You're right.  I wasn't thinking about updating PVs but more what kind of pods it requires (none at present).

The binder's permissions are similar to the provisioner's.

Comment (pweil-) [inline pkg/cmd/server/start/start_master.go]: cool, in that case it sounds like a quick copy/paste of the existing role, removing the extra permissions it doesn't need, and using that new SA/role to get the client.

**Comment by markturansky** [comment]
I removed the special client from the provisioner.  It's now exactly like binder and the rest of the controllers that do not make pods.

**Comment by markturansky** [comment]
@deads2k per our discussion, i added 3 distinct SAs for binder/recycler/provisioner.

**Comment by markturansky** [comment]
@deads2k If the SAs match what is expected, can we begin the merge process?   The EBS provisioner just got merged upstream and the patch PR (https://github.com/openshift/origin/pull/6308) requires this PR to be merged first.

**Comment by deads2k** [comment]
The SA stuff is structurally correct, I didn't inspect the actual permissions.

I didn't review the upstream code.

**Comment by markturansky** [comment]
Upstream was put through the wringer and merged.  Review time for that has passed, but it got a lof RH attention w/ final from Google.

We need to get this PR merged so we can next merge the provisioners.  This is 3.1.1 functionality.

**Comment by ncdc** [comment]
This PR replaces the service account "pv-controller" with 3 new ones. Are there any upgrade implications, moving from 3.1 to 3.1.1?

**Comment by liggitt** [comment]
new ones get created as needed

**Comment by liggitt** [comment]
though sccs will need to be reconciled (they aren't automatically reconciled during upgrade today...)

**Comment by markturansky** [comment]
The code that retrieves the client for a specific SA uses the constant value.  I don't know if SAs are changeable by an admin, but the recycler would break if it were renamed.

**Comment by ncdc** [comment]
@liggitt so the only thing that an admin needs to do is reconcile SCCs manually, and otherwise, everything should just work? If so, can we m-e-r-g-e?

**Comment by liggitt** [comment]
@ncdc yes

**Comment by ncdc** [comment]
[merge]

**Comment by liggitt** [comment]
need something to track the reconcile (either in doc or in ansible, though we don't have a good way to compute "covers" for an SCC, so I'm wary of auto-reconciling until we do)

**Comment by ncdc** [comment]
@sdodson @detiber @brenton @adellape FYI, when upgrading from 3.1 to 3.1.1, there is some SCC reconciliation that will be required because of this PR. How do you want to handle making sure admins know about it?

**Comment by sdodson** [comment]
Is it handled by the standard policy reconciliation that we request users
do during all upgrades?
https://docs.openshift.com/enterprise/3.1/install_config/upgrades.html#updating-policy-definitions

If not we need to call it out as a version specific requirement for 3.1.1
and 1.1.1

**Comment by liggitt** [comment]
it is distinct. reviewing and reconciling should be done on every upgrade, imo.

**Comment by markturansky** [comment]
@sdodson it seems like we will have to call it out as an upgrade step due to the SA name change.

**Comment by sdodson** [comment]
> @sdodson it seems like we will have to call it out as an upgrade step due to the SA name change.

Mind filing either an issue or PR against https://github.com/openshift/openshift-docs with rough outline of what someone upgrading would need to do? Tag me encase we can come up with a way to automate it.

**Comment by liggitt** [comment]
the SA gets created automatically. the `oadm policy reconcile-sccs` command needs to be run to ensure the new SAs have the right level of access for creating pods

**Comment by pweil-** [comment]
current docs on updating SCCs https://docs.openshift.org/latest/admin_guide/manage_scc.html#updating-security-context-constraints

**Comment by markturansky** [comment]
[test]

**Comment by sdodson** [comment]
Is `oadm policy reconcile-sccs --confirm` a sane thing to do during all
upgrades followed up with `oadm policy reconcile-scc --additive-only=false`
to alert admins to things they may want to opt into?

**Comment by liggitt** [comment]
depends on how strongly we claim ownership of the sccs we define. we don't have a good way to compute "covers" for an SCC, so I'm wary of auto-reconciling until we do. `--additive-only` preserves users, groups, and priorities they have set, but overwrites the scc permissions

**Comment by markturansky** [comment]
All tests passed.

**Comment by ncdc** [comment]
[merge]

**Comment by pmorie** [comment]
@markturansky Looks like test-cmd.sh failed:

```

2015-12-17 12:33:11.209686 I | profile: cpu profiling enabled, cpu.pprof
2015-12-17 12:33:11.266835 I | profile: cpu profiling enabled, cpu.pprof
2015-12-17 12:33:35.646059 I | profile: cpu profiling enabled, cpu.pprof
2015-12-17 12:33:43.852858 I | profile: cpu profiling enabled, cpu.pprof
2015-12-17 12:33:43.906608 I | profile: cpu profiling enabled, cpu.pprof
Error from server: User "system:anonymous" cannot list pods in project "default"
Dry run enabled - no modifications will be made. Add --confirm to remove images
!!! Error in hack/../test/cmd/../../hack/cmd_util.sh:407
	'return 1' exited with status 1
Call stack:
	1: hack/../test/cmd/../../hack/cmd_util.sh:407 os::cmd::try_until_success(...)
	2: hack/../test/cmd/newapp.sh:21 main(...)
Exiting with status 1
!!! Error in hack/test-cmd.sh:284
	'${test}' exited with status 1
Call stack:
	1: hack/test-cmd.sh:284 main(...)
Exiting with status 1
```

**Comment by markturansky** [comment]
@pmorie yes, but I think that is an unrelated issue.

**Comment by pmorie** [comment]
[merge]

**Comment by ncdc** [comment]
@pmorie it's already in the merge queue :-p

**Comment by pmorie** [comment]
@ncdc my bad, i had thought we you had to re-tag to re-trigger the tests.  Mark says the test error is a flake.

**Comment by ncdc** [comment]
Woot

**Comment by markturansky** [comment]
Progress!  Thanks, everyone.

**Pull request merged**